### ISSUE TITLE
[Chapter 17] Unsupervised pretraining: Exercise Solutions: Model cloning error

### ISSUE THREAD

**Issue by sarajcev** [ageron/handson-ml2#246]
In the solutions to the exercises at the end of the notebook [17_autoencoders_and_gans.ipynb], there is the example of an unsupervised pre-training using the autoencoder [Cell: 69]. The point of the exercise was to reuse the _encoder_ part of the autoencoder (that was trained on "the large set of supposedly unlabeled data") in order to copy/transfer its parameters (similar to the transfer learning approach) into a new classifier (which is trained on a small set of labeled data). 

The encoder was cloned `conv_encoder_clone = keras.models.clone_model(conv_encoder)` but the layer weighs were **not** copied/transferred? According to the TF2 documentation, model cloning using `clone_model` does **not** transfer weights.

**Question**: Is there a need to transfer weights at this point between the original model and its clone? In other words, is the following line of code `conv_encoder_clone.set_weights(conv_encoder.get_weights())` missing from the Cell 69?

In addition, when the model `pretrained_clf` (i.e. the new classifier that was built on top of the encoder) is un-frozen for the final fine-tuning, should the `learning_rate` be _lowered_ [Cell: 71] and not left as it was `keras.optimizers.SGD(lr=0.02)`?

**Comment by uzi0espil** [comment]
Yes I totally agree with you, the clone re-initialize the weights and does not transfer the learned representation.. I would add the following:

~~~python
conv_encoder_clone= tf.keras.models.clone_model(conv_encoder)
for i, layer in enumerate(conv_encoder.layers[1:]):  # we drop the Reshape layer
    conv_encoder_clone.layers[i].set_weights(layer.get_weights())
    conv_encoder_clone.layers[i].trainable = False  # freeze layer
~~~

Regarding the unfreezing part, I agree, the learning rate should be reduced in order not to wreck the already trained weights.

**Comment by sarajcev** [comment]
@uzi0espil thank you for your comments. I could open a PR and address these issues, if the owner/ maintainer(s) would confirm this is the way to proceed? I appreciate all the work that went into producing these notebooks, as well as the fact that they are offered freely, and would like to contribute back. Thanks!

**Comment by ageron** [comment]
Hi @sarajcev ,

Thanks for your feedback. You're absolutely right, I forgot the `model.set_weights()` line, and it would have been a good idea to reduce the learning rate after unfreezing the pretrained part of the model.

@uzi0espil, your code works fine, thanks! That said, `conv_encoder_clone.set_weights(conv_encoder.get_weights())` followed by `conv_encoder_clone.trainable = False` would be a bit simpler and produce the same result, I believe.

That said, while investigating this issue, I realized that the whole section is actually old code that I forgot to remove from the notebook! This is why there were two "Exercise solutions" sections, and the first one was completely outdated. Oops... I'm sorry about that. I just removed the first one, which contained the outdated code.

However I left the part about semantic hashing, as I think it's interesting, even though there are much more advanced models out there, it gives an idea of what it's about. I added lot of comments, so hopefully it will be clear to everyone.

@sarajcev , thanks as well for your kind words and your offer to submit a PR, I really appreciate it. If you would like to contribute to this project, it would be absolutely great if you could add the solution to exercises 10 or 11 in chapter 17, or perhaps the solution to exercise 8 in chapter 18 if you prefer: these are the last remaining exercises without a solution. That would be very helpful! 😊
